File: 17. RotateBits.cpp
// Topic: Rotate bits left and right by k positions for given bit-width
#include <bits/stdc++.h>
using namespace std;

unsigned int rotateLeftN(unsigned int n, int d, int bits)
{
    d %= bits;
    return ((n << d) & ((1u << bits) - 1)) | (n >> (bits - d));
}

unsigned int rotateRightN(unsigned int n, int d, int bits)
{
    d %= bits;
    return (n >> d) | ((n << (bits - d)) & ((1u << bits) - 1));
}

int main()
{
    unsigned int x = 10;
    cout << "RotateLeft 10 by2 (8bits): " << rotateLeftN(x, 2, 8) << '\n';
    cout << "RotateRight 10 by2 (8bits): " << rotateRightN(x, 2, 8) << '\n';
    return 0;
}

// Time Complexity: O(1)
// Space Complexity: O(1)

File: 17. RotateBits.py
# Topic: Rotate bits left and right by k positions for given bit-width

def rotate_left_n(n, d, bits=32):
    d %= bits
    return ((n << d) & ((1<<bits)-1)) | (n >> (bits - d))

def rotate_right_n(n, d, bits=32):
    d %= bits
    return (n >> d) | ((n << (bits - d)) & ((1<<bits)-1))

if __name__ == '__main__':
    print('RotateLeft 10 by2 (8bits):', rotate_left_n(10,2,8))
    print('RotateRight 10 by2 (8bits):', rotate_right_n(10,2,8))

# Time Complexity: O(1)
# Space Complexity: O(1)
